Visualize Frustum

BA-Utkarsh opened the pull request:
# 🎉 New feature

## Summary
This PR mainly adds the visualization of Frustum.
We could see it was present in gazebo classic and from gazebo garden onwards the plugin/feature is not available.

## Test it
$ Build gazebo from source.
$ . install/setup.sh
$ gz sim examples/worlds/visualize_frustum.sdf

## Test Ref images,
1. Play the simulation.
![frustum-1](https://github.com/user-attachments/assets/fef2b73b-e838-4d59-aa06-9c725fbef2ba)

2. Select the topic from scroll down.
![frustum-2](https://github.com/user-attachments/assets/df28d6c7-6f7e-4379-b123-8906a5c1809c)

3. Refresh it to get the "logical_camera/frustum" topic.
![frustum-3](https://github.com/user-attachments/assets/ae43d9c8-6c57-43d0-9420-d11c3280d277)

4. Subcriibed to "logical_camera/frustum".
![frustum-4](https://github.com/user-attachments/assets/70cf896e-56f7-4604-993b-a0d11a48fd68)

5. Final output
![frustum-6](https://github.com/user-attachments/assets/83272129-bc1a-4d97-a8c9-d3883d5c1bae)

## Checklist
- [x] Signed all commits for DCO
- [x] Added tests
- [ ] Added example and/or tutorial
- [ ] Updated documentation (as needed)
- [ ] Updated migration guide (as needed)
- [ ] Consider updating Python bindings (if the library has them)
- [ ] `codecheck` passed
- [ ] All tests passed
- [ ] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Agazebosim+archived%3Afalse+) to support the maintainers

## Supporting PRs
- https://github.com/gazebosim/gz-sim/pull/2707
- https://github.com/gazebosim/gz-sensors/pull/491

ahcorde commented on include/gz/rendering/FrustumVisual.hh:
```diff
@@ -0,0 +1,180 @@
+/*
+ * Copyright (C) 2015 Open Source Robotics Foundation
```
```suggestion
 * Copyright (C) 2024 Open Source Robotics Foundation
```

BA-Utkarsh replied:
Updated!

ahcorde commented on include/gz/rendering/FrustumVisual.hh:
```diff
@@ -0,0 +1,180 @@
+      /// \brief Destructor
+      public: virtual ~FrustumVisual();
+
+      /// \brief Constructor
+      /// \param[in] _near Near plane distance. This is the distance from
+      /// the frustum's vertex to the closest plane
+      /// \param[in] _far Far plane distance. This is the distance from the
+      /// frustum's vertex to the farthest plane.
+      /// \param[in] _fov Field of view. The field of view is the
+      /// angle between the frustum's vertex and the edges of the near or far
+      /// plane. This value represents the horizontal angle.
+      /// \param[in] _aspectRatio The aspect ratio, which is the width divided
+      /// by height of the near or far planes.
+      /// \param[in] _pose Pose of the frustum, which is the vertex (top of
+      /// the pyramid).
+      /*protected: FrustumVisual(double _near,
+                      double _far,
+                      const math::Angle &_fov,
+                      double _aspectRatio,
+                      const gz::math::Pose3d &_pose = gz::math::Pose3d::Zero);*/
```
remove ?

BA-Utkarsh replied:
Updated!

ahcorde commented on include/gz/rendering/FrustumVisual.hh:
```diff
@@ -0,0 +1,180 @@
+      /// \return Plane of the frustum.
+      public: virtual gz::math::Planed Plane(const FrustumVisualPlane _plane) const = 0;
+
+      /// \brief Check if a box lies inside the pyramid frustum.
+      /// \param[in] _b Box to check.
+      /// \return True if the box is inside the pyramid frustum.
+      //public: virtual bool Contains(const gz::math::AxisAlignedBox &_b) const = 0; //TO-DO
+
+      /// \brief Check if a point lies inside the pyramid frustum.
+      /// \param[in] _p Point to check.
+      /// \return True if the point is inside the pyramid frustum.
+      //public: virtual bool Contains(const gz::math::Vector3d &_p) const = 0; //TO-DO
```
remove ?

BA-Utkarsh replied:
Updated!

ahcorde commented on include/gz/rendering/FrustumVisual.hh:
```diff
@@ -0,0 +1,180 @@
+      private: void ComputePlanes();
+
+      /// \brief Private data pointer
+      //GZ_UTILS_IMPL_PTR(dataPtr)
```
remove ?

BA-Utkarsh replied:
Updated!

ahcorde commented on include/gz/rendering/base/BaseFrustumVisual.hh:
```diff
@@ -0,0 +1,265 @@
+/*
+ * Copyright (C) 2020 Open Source Robotics Foundation
```
```suggestion
 * Copyright (C) 2024 Open Source Robotics Foundation
```

BA-Utkarsh replied:
Updated!

ahcorde commented on include/gz/rendering/base/BaseFrustumVisual.hh:
```diff
@@ -0,0 +1,265 @@
+      protected: double aspectRatio = 1.0;
+
+      /// \brief array of plane
+      protected: std::array<gz::math::Planed, 6> planes;
```
include `<array>`

ahcorde commented on include/gz/rendering/base/BaseFrustumVisual.hh:
```diff
@@ -0,0 +1,265 @@
+    template <class T>
+    gz::math::Pose3d BaseFrustumVisual<T>::Pose() const
+    {
+	    return this->pose;
```
```suggestion
      return this->pose;
```

ahcorde commented on include/gz/rendering/base/BaseFrustumVisual.hh:
```diff
@@ -0,0 +1,265 @@
+        mtl->SetMetalness(0.1f);
+        mtl->SetReflectivity(0.2);
+      }
+      return;
```
remove

ahcorde commented on include/gz/rendering/base/BaseFrustumVisual.hh:
```diff
@@ -0,0 +1,265 @@
+#ifndef GZ_RENDERING_BASEFRUSTUMVISUAL_HH_
+#define GZ_RENDERING_BASEFRUSTUMVISUAL_HH_
+
+#include <vector>
```
is vector used here ? I can see `std::array`

ahcorde commented on ogre/include/gz/rendering/ogre/OgreFrustumVisual.hh:
```diff
@@ -0,0 +1,77 @@
+/*
+ * Copyright (C) 2020 Open Source Robotics Foundation
```
```suggestion
 * Copyright (C) 2024 Open Source Robotics Foundation
```

ahcorde commented on ogre/include/gz/rendering/ogre/OgreFrustumVisual.hh:
```diff
@@ -0,0 +1,77 @@
+#ifndef GZ_RENDERING_OGRE_OGREFRUSTUMVISUAL_HH_
+#define GZ_RENDERING_OGRE_OGREFRUSTUMVISUAL_HH_
+
+#include <vector>
```
```suggestion
```

ahcorde commented on ogre/include/gz/rendering/ogre/OgreFrustumVisual.hh:
```diff
@@ -0,0 +1,77 @@
+#define GZ_RENDERING_OGRE_OGREFRUSTUMVISUAL_HH_
+
+#include <vector>
+#include <memory>
+#include "gz/rendering/base/BaseFrustumVisual.hh"
```
```suggestion
#include <memory>

#include "gz/rendering/base/BaseFrustumVisual.hh"
```

ahcorde commented on ogre/src/OgreFrustumVisual.cc:
```diff
@@ -0,0 +1,102 @@
+/*
+ * Copyright (C) 2020 Open Source Robotics Foundation
```
```suggestion
 * Copyright (C) 2024 Open Source Robotics Foundation
```

ahcorde commented on ogre/src/OgreFrustumVisual.cc:
```diff
@@ -0,0 +1,102 @@
+  /// \brief Frustum Ray DynamicLines Object to display
+  public: std::vector<std::shared_ptr<OgreDynamicLines>> rayLines;
+
+  /// \brief Frustum visual type
+  //public: FrustumVisualPlane frustumVisPlane =
+  	  //FrustumVisualPlane::FRUSTUM_PLANE_TOP;
```
remove

ahcorde commented on ogre/src/OgreFrustumVisual.cc:
```diff
@@ -0,0 +1,102 @@
+  public: bool visible = true;
+
+  /// \brief Each corner of the frustum.
+  public: std::array<gz::math::Vector3d, 8> points;
```
include `<array>`

iche033 replied:
[19aa075](https://github.com/gazebosim/gz-rendering/pull/1095/commits/19aa07588cd84d2e56e5cfc2435b4a4772671642)

ahcorde commented on ogre2/include/gz/rendering/ogre2/Ogre2FrustumVisual.hh:
```diff
@@ -0,0 +1,77 @@
+/*
+ * Copyright (C) 2020 Open Source Robotics Foundation
```
```suggestion
 * Copyright (C) 2024 Open Source Robotics Foundation
```

ahcorde commented on ogre2/include/gz/rendering/ogre2/Ogre2FrustumVisual.hh:
```diff
@@ -0,0 +1,77 @@
+#define GZ_RENDERING_OGRE2_OGREFRUSTUMVISUAL_HH_
+
+#include <memory>
+#include <vector>
```
```suggestion

```

ahcorde commented on ogre2/src/Ogre2FrustumVisual.cc:
```diff
@@ -0,0 +1,290 @@
+/*
+ * Copyright (C) 2020 Open Source Robotics Foundation
```
```suggestion
 * Copyright (C) 2024 Open Source Robotics Foundation
```

ahcorde commented on ogre2/src/Ogre2FrustumVisual.cc:
```diff
@@ -0,0 +1,290 @@
+  /// \brief Frustum Ray DynamicLines Object to display
+  public: std::vector<std::shared_ptr<Ogre2DynamicRenderable>> rayLines;
+
+  /// \brief Frustum visual type
+  //public: FrustumVisualPlane frustumVisPlane =
+  	  //FrustumVisualPlane::FRUSTUM_PLANE_TOP;
```
remove

ahcorde commented on ogre2/src/Ogre2FrustumVisual.cc:
```diff
@@ -0,0 +1,290 @@
+class gz::rendering::Ogre2FrustumVisualPrivate
+{
+  /// \brief Frustum Ray DynamicLines Object to display
+  public: std::vector<std::shared_ptr<Ogre2DynamicRenderable>> rayLines;
```
include `<vector>`

iche033 replied:
[19aa075](https://github.com/gazebosim/gz-rendering/pull/1095/commits/19aa07588cd84d2e56e5cfc2435b4a4772671642)

ahcorde commented on ogre2/src/Ogre2FrustumVisual.cc:
```diff
@@ -0,0 +1,290 @@
+  public: bool visible = true;
+
+  /// \brief Each corner of the frustum.
+  public: std::array<gz::math::Vector3d, 8> points;
```
include `<array>`

BA-Utkarsh replied:
Updated!

ahcorde commented on ogre2/src/Ogre2FrustumVisual.cc:
```diff
@@ -0,0 +1,290 @@
+{
+  // enable GL_PROGRAM_POINT_SIZE so we can set gl_PointSize in vertex shader
+  auto engine = Ogre2RenderEngine::Instance();
+  std::string renderSystemName =
```
include `<string>`

BA-Utkarsh replied:
Updated!

ahcorde commented on ogre2/src/Ogre2FrustumVisual.cc:
```diff
@@ -0,0 +1,290 @@
+//////////////////////////////////////////////////
+void Ogre2FrustumVisual::Update()
+{
+  std::shared_ptr<Ogre2DynamicRenderable> renderable =
```
include `<memory>`

BA-Utkarsh replied:
Updated!

ahcorde commented on ogre2/src/Ogre2FrustumVisual.cc:
```diff
@@ -0,0 +1,290 @@
+  this->dataPtr->rayLines.push_back(renderable);
+
+  // Tangent of half the field of view.
+  double tanFOV2 = std::tan(this->fov() * 0.5);
```
include `<cmath>`

BA-Utkarsh replied:
Updated!

ahcorde commented on src/FrustumVisual.cc:
```diff
@@ -0,0 +1,28 @@
+/*
+ * Copyright (C) 2020 Open Source Robotics Foundation
```
```suggestion
 * Copyright (C) 2024 Open Source Robotics Foundation
```

BA-Utkarsh replied:
Updated!

BA-Utkarsh commented:
@ahcorde , thank you for review. I have addressed all the review comments.

ahcorde reviewed (changes requested):
please check [linters](https://github.com/gazebosim/gz-rendering/actions/runs/12514292293/job/35270138258?pr=1095):

```
  /usr/share/gz/gz-cmake4/codecheck/cpplint.py:55: DeprecationWarning: module 'sre_compile' is deprecated
    import sre_compile
  /github/workspace/include/gz/rendering/FrustumVisual.hh:131:  Lines should be <= 80 characters long  [whitespace/line_length] [2]
  /github/workspace/include/gz/rendering/Scene.hh:1121:  Lines should be <= 80 characters long  [whitespace/line_length] [2]
  /github/workspace/include/gz/rendering/base/BaseFrustumVisual.hh:81:  Lines should be <= 80 characters long  [whitespace/line_length] [2]
  /github/workspace/include/gz/rendering/base/BaseFrustumVisual.hh:84:  Lines should be <= 80 characters long  [whitespace/line_length] [2]
  /github/workspace/include/gz/rendering/base/BaseFrustumVisual.hh:84:  Should have a space between // and comment  [whitespace/comments] [4]
  /github/workspace/include/gz/rendering/base/BaseFrustumVisual.hh:87:  Lines should be <= 80 characters long  [whitespace/line_length] [2]
  /github/workspace/include/gz/rendering/base/BaseFrustumVisual.hh:87:  Should have a space between // and comment  [whitespace/comments] [4]
  /github/workspace/include/gz/rendering/base/BaseFrustumVisual.hh:222:  Lines should be <= 80 characters long  [whitespace/line_length] [2]
  /github/workspace/include/gz/rendering/base/BaseFrustumVisual.hh:238:  Tab found; better to use spaces  [whitespace/tab] [1]
  /github/workspace/include/gz/rendering/base/BaseScene.hh:850:  Lines should be <= 80 characters long  [whitespace/line_length] [2]
  /github/workspace/ogre/include/gz/rendering/ogre/OgreScene.hh:184:  Lines should be <= 80 characters long  [whitespace/line_length] [2]
  /github/workspace/ogre/src/OgreFrustumVisual.cc:32:  Should have a space between // and comment  [whitespace/comments] [4]
  /github/workspace/ogre/src/OgreFrustumVisual.cc:33:  Tab found; better to use spaces  [whitespace/tab] [1]
  /github/workspace/ogre/src/OgreFrustumVisual.cc:33:  Should have a space between // and comment  [whitespace/comments] [4]
  /github/workspace/ogre/src/OgreFrustumVisual.cc:42:  Lines should be <= 80 characters long  [whitespace/line_length] [2]
  /github/workspace/ogre/src/OgreFrustumVisual.cc:94:  Should have a space between // and comment  [whitespace/comments] [4]
  /github/workspace/ogre2/include/gz/rendering/ogre2/Ogre2Scene.hh:347:  Lines should be <= 80 characters long  [whitespace/line_length] [2]
  /github/workspace/ogre2/src/Ogre2FrustumVisual.cc:61:  Should have a space between // and comment  [whitespace/comments] [4]
  /github/workspace/ogre2/src/Ogre2FrustumVisual.cc:62:  Tab found; better to use spaces  [whitespace/tab] [1]
  /github/workspace/ogre2/src/Ogre2FrustumVisual.cc:62:  Should have a space between // and comment  [whitespace/comments] [4]
  /github/workspace/ogre2/src/Ogre2FrustumVisual.cc:71:  Lines should be <= 80 characters long  [whitespace/line_length] [2]
  /github/workspace/ogre2/src/Ogre2FrustumVisual.cc:151:  Should have a space between // and comment  [whitespace/comments] [4]
  /github/workspace/ogre2/src/Ogre2FrustumVisual.cc:154:  Line ends in whitespace.  Consider deleting these extra spaces.  [whitespace/end_of_line] [4]
  /github/workspace/ogre2/src/Ogre2FrustumVisual.cc:175:  Lines should be <= 80 characters long  [whitespace/line_length] [2]
  /github/workspace/ogre2/src/Ogre2FrustumVisual.cc:176:  Lines should be <= 80 characters long  [whitespace/line_length] [2]
  /github/workspace/ogre2/src/Ogre2FrustumVisual.cc:177:  Lines should be <= 80 characters long  [whitespace/line_length] [2]
  /github/workspace/ogre2/src/Ogre2FrustumVisual.cc:193:  Lines should be <= 80 characters long  [whitespace/line_length] [2]
  /github/workspace/ogre2/src/Ogre2FrustumVisual.cc:194:  Lines should be <= 80 characters long  [whitespace/line_length] [2]
  /github/workspace/ogre2/src/Ogre2FrustumVisual.cc:195:  Lines should be <= 80 characters long  [whitespace/line_length] [2]
  /github/workspace/ogre2/src/Ogre2FrustumVisual.cc:268:  Lines should be <= 80 characters long  [whitespace/line_length] [2]
  /github/workspace/ogre2/src/Ogre2FrustumVisual.cc:269:  Lines should be <= 80 characters long  [whitespace/line_length] [2]
  /github/workspace/ogre2/src/Ogre2FrustumVisual.cc:272:  Lines should be <= 80 characters long  [whitespace/line_length] [2]
  /github/workspace/ogre2/src/Ogre2FrustumVisual.cc:275:  Lines should be <= 80 characters long  [whitespace/line_length] [2]
  /github/workspace/ogre2/src/Ogre2FrustumVisual.cc:278:  Lines should be <= 80 characters long  [whitespace/line_length] [2]
  /github/workspace/ogre2/src/Ogre2FrustumVisual.cc:281:  Lines should be <= 80 characters long  [whitespace/line_length] [2]
  /github/workspace/ogre2/src/Ogre2FrustumVisual.cc:283:  Lines should be <= 80 characters long  [whitespace/line_length] [2]
  /github/workspace/ogre2/src/Ogre2FrustumVisual.cc:284:  Lines should be <= 80 characters long  [whitespace/line_length] [2]
  Total errors found: 37
```

iche033 commented:
Thanks for the contribution!

We follow semantic versioning and promise to maintain ABI compatibility within the same major version. Since this is targeting a release branch, we can't break ABI. The [ABI checker](https://build.osrfoundation.org/job/gz_rendering-abichecker-any_to_any-ubuntu-noble-amd64/64/API_5fABI_20report/) did pick up that the changes broke ABI (compatibility needs to be 100%). The breaking change is mainly the new `CreateFrustumVisualImpl` virtual function (which changed the vtable for BaseScene class).

Ideally we should re-target this PR to the `main` branch. Do you need this is `gz-rendering9`? If I not, let's just re-target to `main`. If you do need this in a released branch (gz-rendering9 or earlier), I can help implement a workaround to make it ABI-compatible. To do this, we should also get this in `main` first, and I'll backport this change with ABI compatibility changes.

BA-Utkarsh commented:
> Thanks for the contribution!
> 
> We follow semantic versioning and promise to maintain ABI compatibility within the same major version. Since this is targeting a release branch, we can't break ABI. The [ABI checker](https://build.osrfoundation.org/job/gz_rendering-abichecker-any_to_any-ubuntu-noble-amd64/64/API_5fABI_20report/) did pick up that the changes broke ABI (compatibility needs to be 100%). The breaking change is mainly the new `CreateFrustumVisualImpl` virtual function (which changed the vtable for BaseScene class).
> 
> Ideally we should re-target this PR to the `main` branch. Do you need this is `gz-rendering9`? If I not, let's just re-target to `main`. If you do need this in a released branch (gz-rendering9 or earlier), I can help implement a workaround to make it ABI-compatible. To do this, we should also get this in `main` first, and I'll backport this change with ABI compatibility changes.

Thank you @iche033 for your comment and review.
Yes, I do need it in gz-rendering9.
I shall change the target to `main`. Please help me with the ABI compatibility.

BA-Utkarsh commented:
@iche033 , I changed the target branch to `main`, please let me know the next action items or you will take care from here? :)

iche033 commented:
I created a PR (https://github.com/BA-Utkarsh/gz-rendering/pull/2) to your repo with some changes to function and variable names for consistency with the Camera class. I also added a `frustum_visual` demo, and addressed comments from other reviewers. Please take a look and merge if changes look good to you.

BA-Utkarsh commented:
> I created a PR ([BA-Utkarsh#2](https://github.com/BA-Utkarsh/gz-rendering/pull/2)) to your repo with some changes to function and variable names for consistency with the Camera class. I also added a `frustum_visual` demo, and addressed comments from other reviewers. Please take a look and merge if changes look good to you.

@iche033 , I have merged it.
